### PR TITLE
feat(openai): add a request-head operation classifier

### LINKED ISSUE #744: feat(ai): add a request-head operation classifier
## Description

Add an `ai_operation` filter that classifies supported application operations
from the HTTP request head and publishes a provider-neutral typed match for
downstream filters.

The classifier aggregates protocol-owned registries through the shared matcher.
It does not inspect request bodies or infer the provider that will ultimately
serve the request.

Parent epic: #740

## Scope

- Run classification during `on_request` with no request-body access.
- Match using the provider-neutral operation registry foundation.
- Aggregate registered OpenAI, Anthropic, and future application protocols
  without protocol-specific branches in the classifier.
- Insert a typed `AiOperationMatch` into request extensions.
- Publish stable metadata and filter results:
  - `ai_operation.application_protocol`
  - `ai_operation.operation_id`
- Support configurable proxy-owned routing headers:
  - `x-praxis-ai-application-protocol`
  - `x-praxis-ai-operation`
- Remove or overwrite client-supplied values for configured routing headers.
- Leave unmatched requests unchanged and let routing policy determine fallback
  behavior.
- Register the filter and provide a focused example configuration.
- Document that current `on_request` publication follows `StreamBuffer`
  pre-read; early visibility to body processors is tracked by
  praxis-proxy/praxis#1142.

## Acceptance Criteria

- The filter declares no request-body access or buffering mode.
- Supported method/path/transport combinations publish the expected typed match,
  metadata, results, and configured headers.
- Unsupported methods and paths publish no operation match.
- Adding a protocol registry requires no protocol-specific branch in
  `ai_operation`.
- OpenAI Responses HTTP and WebSocket operations remain distinguishable.
- Client-supplied operation headers cannot influence the authoritative match.
- No `application_provider` is inferred or published from the incoming request.
- Unit tests, functional integration tests, example configuration, generated
  example inventory, and filter documentation are complete.

## Dependencies

- #1050 — provider-neutral operation identity and matcher

## Related Work

- praxis-proxy/praxis#1142 — make this match available before `StreamBuffer`
  pre-read
- praxis-proxy/praxis#1075 — router visibility of pending header mutations

## Area

Provider APIs (OpenAI/Anthropic)

### PR DESCRIPTION
**Pull request by cdoern**
Closes #744

## Stacked on #788

The first commit, `e209fa8f`, is the #743 Responses registry under review in #788.
It is included because the classifier consults that registry. **Merge #788 first**;
this branch then contains only its own commit. Review just `35951360` here.

## Summary

Operation identity was derived per filter from path string matching, so every
consumer re-implemented the same method and path checks and no single component
owned the answer.

`openai_operation` classifies supported operations from the request head alone
and publishes the result four ways:

| Output | Consumer |
| --- | --- |
| typed `OpenAiOperationMatch` in request extensions | downstream filters (#742, #741) |
| `openai_operation.family`, `openai_operation.operation_id` metadata | logging, tracing |
| filter results `family`, `operation_id` | `on_result` branch conditions |
| `x-praxis-ai-family`, `x-praxis-ai-operation` headers | see the note below |

The filter takes the default `BodyAccess::None` and `BodyMode::Stream`, so it
declares no body access and no buffering — the classification is available
before any body-handling decision is made.

Transport comes from the opening handshake headers (RFC 6455 §4.1, with
`Connection` treated as a token list per RFC 9110 §7.6.1). That is what separates
the Responses WebSocket operation from `POST /v1/responses` at the same path
without inspecting a payload.

## Note on consuming the classification

Two findings from wiring this up, both worth knowing before #745 starts.

**`x-praxis-ai-*` is a reserved prefix.** The protocol layer rejects a client
request carrying one with a 400, and strips them before forwarding upstream. So
the "client-supplied headers cannot spoof the classifier" criterion is already
enforced at the ingress boundary, ahead of any filter. This filter's
overwrite-on-match and remove-on-miss remain as defense in depth.

**Header-phase mutations are not visible to `router`.** The router matches the
downstream request headers, while a header-phase filter's changes are pending
until the request is forwarded. `openai_responses_format` only routes on
`x-praxis-ai-format` because it runs in the pre-read body phase, which lands
before the header-phase pipeline. A head-only classifier has no equivalent.

Consequently the published headers have no in-pipeline consumer today. The
working mechanisms are request extensions for downstream filters and filter
results for branch chains. The header code is kept because it is correct, cheap,
and becomes useful with a non-reserved header name or a future consumer — but
#745's plan to route on these headers needs revisiting before it is picked up.

## Testing

- `cargo test --workspace` — all suites pass, including 2594 in `praxis-ai-apis`
  and 479 integration tests
- `make lint` — passes end to end
- `make doc` — clean

15 unit tests cover the published outputs: both families classified, WebSocket
separated from create by transport, static endpoints not consumed as
identifiers, unsupported methods publishing nothing, configurable and disabled
header names, invalid header names rejected at startup, and the
`Connection`/`Upgrade` token-list edge cases.

6 functional tests cover the example config's observable boundary behaviour:
classified and unclassified traffic forwarded unharmed, the classifier's own
headers never crossing to upstream, and client-supplied reserved headers
rejected at ingress on both classified and unclassified paths.

## Not in scope

Switching the existing classifier and Responses filters to consume this match is
#742 and #741. Registry-driven routing is #745.

### CODE REVIEW

File: docs/filters/openai_operation.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<!-- Generated by: cargo xtask generate-filter-docs -->
+<!-- Do not edit manually -->
+
+# `openai_operation`
+
+One operation classified from a request head.
+
```
[the remark this answers is not in the document]
Comment (leseb) [reply docs/filters/openai_operation.md]: https://github.com/praxis-proxy/ai/pull/890/commits/60e77fb56b5a185159c75d1a916511f7c9495bef

File: apis/src/openai/operation_classifier/config.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+///
+/// Both headers are proxy-owned. A configured name is always either
+/// overwritten with the classifier's own value or removed, so a client
+/// cannot supply one and influence routing.
+#[derive(Debug, Clone, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct OperationHeaders {
+    /// Header name for the API family. `null` disables the header.
+    #[serde(default = "default_family_header")]
+    pub family: Option<String>,
```
Comment (leseb) [inline apis/src/openai/operation_classifier/config.rs:25]: let's use application_protocol instead so this matches https://github.com/praxis-proxy/praxis/pull/1106/ - sorry the change, just realized now this would be better

**Review by leseb — CHANGES_REQUESTED** [review]
P2 — Unsupported upgrade headers suppress HTTP classification. mod.rs:118 selects WebSocket transport without requiring GET. Adding upgrade headers to a normal operation such as POST /v1/responses leaves it unclassified. This is currently latent because WebSocket and classifier-based policy routing are not supported yet, but should be fixed before consumers land.

P2 — Unsafe header targets are accepted. config.rs:87 permits targets such as authorization, host, and content-length, potentially overwriting or removing authentication and transport state. Use the repository’s shared promotion-header validation.

P2 — Duplicate output headers are accepted. The same validation permits family and operation to use the same header name. The operation value then overwrites the family value. Reject duplicates case-insensitively.

P2 — Functional tests do not observe classification. operation_classifier.rs:47 tests forwarding and core reserved-header handling, but never consumes the classifier result. The suite would pass with a no-op classifier. Route distinct operations through on_result to distinct backends and assert the selected backend.

**Review by leseb — APPROVED** [review]
Please do this as a followup: 

P2 remains: config.rs:138 still permits x-api-key, api-key, x-goog-api-key, and unrelated internal headers such as x-praxis-route. The classifier could overwrite or remove credentials/routing state. It should use the shared promotion-header validator.
